[refactor] Move Logging from `init.d` to Agent

shwetd19 opened the pull request:
## Checklist

- [ ] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [ ] I have manually tested the changes proposed in this pull request.
- [ ] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

#### **Description:**
This PR refactors the logging mechanism in the OpenWISP configuration system to centralize logging in the agent (`openwisp.agent`) instead of directly using the `logger` command in `init.d` scripts.

#### **Changes:**
1. **Removed Direct `logger` Calls:**
   - Removed `logger` calls from `openwisp-reload-config` and `openwisp.init`.

2. **Added Logging to Agent:**
   - Added a `log_message` function in `openwisp.agent` to handle logging.
   - Implemented support for the `--log-message` argument in the agent to log messages.

3. **Updated Scripts:**
   - Updated `openwisp-reload-config` to call the agent for logging via `/usr/sbin/openwisp-config --log-message`.


#### **Testing:**
- Verified that log messages are correctly generated by the agent when:
  - Running `/usr/sbin/openwisp-reload-config`.
  - Using the `--log-message` argument with `/usr/sbin/openwisp-config`.
  - Restarting the OpenWISP service.

#### **Fixes:**
- Resolves #145

okraits commented on openwisp-config/files/openwisp.agent:
```diff
@@ -971,3 +975,10 @@ while true; do
+log_message() {
+	local level="$1"
+	local message="$2"
+	os.execute(string.format('logger -t openwisp -p %s "%s"', level, message))
```
The agent is not written in Python, but in shellscript. Please use `logger` like the existing logging calls.
Besides that, you define 2 arguments for the `log_message` function, but when calling it, you only supply 1 argument. You'll need at least a default value for the second argument, if it is not provided at any time.

shwetd19 replied:
I've implemented your suggested change in my recent commit

okraits commented on openwisp-config/files/openwisp.agent:
```diff
@@ -129,6 +129,10 @@ while [ -n "$1" ]; do
+		--log-message)
+			log_message "$2"
+			exit 0
+			;;
```
I would add another option `--log-level` here to make it possible to pass the log level.
Then I would assign message and level to variables and later call `log_message()` if at least a message is provided.

shwetd19 replied:
added it here : https://github.com/shwetd19/openwisp-config/blob/18fbb4df4b91642a47643ae5c0fc7a85f9a67707/openwisp-config/files/openwisp.agent#L12

okraits commented on openwisp-config/files/openwisp.agent:
```diff
@@ -971,3 +975,15 @@ while true; do
+# Logging function
+log_message() {
+	local level="$1"
+	local message="${2:-}"  # Default to an empty string if the second argument is not provided
```
I would make the message mandatory. For the level I would fall back to `daemon.info` if none is provided.

okraits commented on openwisp-config/files/openwisp.agent:
```diff
@@ -971,3 +975,15 @@ while true; do
+	local level="$1"
+	local message="${2:-}"  # Default to an empty string if the second argument is not provided
+
+	if [ -z "$message" ]; then
```
You don't need this `if` statement. You can provide a default value in the assignment of the `message` variable.

shwetd19 replied:
Why is the `if [ -z "$message" ]` check unnecessary when we already have `local message="${2:-}"`? Could you clarify the reasoning?

devkapilbansal replied:
Hi @shwetd19 
If you see the above comment, Oliver suggested a fall back when message is not provided. In that case, this check will be unnecessary

shwetd19 commented:
Hey @okraits , I've made the suggested changes 

- Added log_message() function in openwisp.agent to handle logging
- Added command-line arguments --log-level and --log-message to agent
- Updated openwisp-reload-config and openwisp.init to use the agent for logging
- Made message parameter mandatory with level defaulting to daemon.info
